[improve][txn] Allow superusers to abort transactions

nicoloboschi opened the pull request:
### Motivation

Super users must be always allowed to abort a transaction even if they're not the original owner.

### Modifications

* Check that only owner or superusers are allowed to perform txn operations (end, add partition and add subscription)

### Verifying this change

- [x] Make sure that the change passes the CI checks.


### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. Please attach the local preview screenshots (run `sh start.sh` at `pulsar/site2/website`) to your PR description, or else your PR might not get merged. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2429,19 +2449,45 @@ protected void handleEndTxn(CommandEndTxn command) {
+                    if (meta.getOwner() != null && !meta.getOwner().equals(role)) {
+                        // allows only owner or superuser
+                        if (service.isAuthenticationEnabled() && service.isAuthorizationEnabled()) {
+                            log.info("check if {} is superuser", role);
```
remove this log

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2421,7 +2429,16 @@ protected void handleAddPartitionToTxn(CommandAddPartitionToTxn command) {
+    }
+
+    private CompletableFuture<Void> failedFutureTxnNotOwned(TxnID txnID) {
+        String msg = String.format(
```
why are you using the logger this way ?

nicoloboschi replied:
it's also in the message of the exception

congbobo184 commented on pulsar-transaction/coordinator/src/main/proto/PulsarTransactionMetadata.proto:
```diff
@@ -51,6 +51,7 @@ message TransactionMetadataEntry {
   optional uint64 start_time      = 9;
   optional uint64 last_modification_time = 10;
   optional uint64 max_local_txn_id = 11;
+  optional string owner          = 12;
```
why not use the `transaction-key` directly?  if we need to implement exclusive transactions, we also need to store the transaction-key in the txn meta. I think to store the role name, and transactionl-key is repeated, and a role will catch multi transaction-key. If users don't use the exclusive transactions, we also can use a default transaction-key to implement it.

nicoloboschi replied:
I've added an answer in the doc's comment. I'm reporting it here

> 
> I'm focusing on "If users don't use the exclusive transactions" since it will be the common use case anyway.
> 
> We can't use the transaction-key as authorization data source because it enforces the admin to allow single users to specific transactions key and adds operational costs.
> From a user's perspective, this is very annoying. 
> 
> Most of the cases would end up in use the default transaction-key (which I don't know how it could be generated). I'm not sure how you could have default keys without impacting other clients in the same namespace and fencing them out unintentionally. 
> 
> 
> Adding in future the transactions auth permission makes sense if you're going to use the per-namespace transactions. This will be based on the current auth model, which is well-known to Pulsar's users.
> 
> 
> Also, for the impact of the storage:
> 1. if we store this "default transaction key" that means you are not using exclusive transactions but you're going to store that anyway.
> 2. Exclusive transactions might have a global flag for enable/disable them. How you're going to deal with authorization then?
> 
> I believe the storage overhead (in case of exclusive transactions) will be acceptable and it helps in making Pulsar more user friendly

congbobo184 replied:
I think we can use the owner name as the default-transaction-key by this owner, in this way, the transaction also can be fenced.
The process is as follows:
1. if the user doesn't specify the transaction key, use the owner's name as transaction-key
2. if the owner's name is the same as the transaction key, the authentication passed
3. if the user specifies the transaction key, the transaction key can correspond to multiple roles

Users can be logged off. the transaction can't log out.

liangyepianzhou commented on pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/MLTransactionMetadataStore.java:
```diff
@@ -303,9 +303,9 @@ public CompletableFuture<Void> addProducedPartitionToTxn(TxnID txnID, List<Strin
-                                        + txnMetaListPair.getLeft().status().name(), e);
+                                log.error("TxnID {} add produced partition error"
+                                                + " with TxnStatus: {}", txnMetaListPair.getLeft().id().toString()
+                                        , txnMetaListPair.getLeft().status().name(), e);
```
Why only store the owner in memory and not store the owner in TransactionMetadataEntry here?

nicoloboschi replied:
it's fixed now, thanks

congbobo184 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2421,7 +2429,16 @@ protected void handleAddPartitionToTxn(CommandAddPartitionToTxn command) {
+                getPrincipal(), txnID
+        );
+        log.warn("[{}] {}", remoteAddress, msg);
+        return CompletableFuture.failedFuture(new RuntimeException(msg));
```
please use `AuthenticationException` or add a new transaction exception

nicoloboschi replied:
I don't think it matters. In this case it's not something you can fix by adding more permissions. It's more like an internal error. The client receives a generic `TransactionCoordinatorClientException` 

I've added a check in the unit test

nicoloboschi committed:
[improve][txn] Allow superusers to abort transactions

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2461,6 +2482,25 @@ protected void handleEndTxn(CommandEndTxn command) {
+                .getTxnMeta(txnID)
+                .thenCompose(meta -> {
+                    final String role = getPrincipal();
+                    if (meta.getOwner() != null && !meta.getOwner().equals(role)) {
```
please add a comment about the case in which meta.owner is null (old versions)

eolivelli reviewed (approved):
+1

excellent

codelipenghui commented on pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/MLTransactionMetadataStore.java:
```diff
@@ -224,50 +227,54 @@ public CompletableFuture<TxnMeta> getTxnMeta(TxnID txnID) {
+                        .setMetadataOp(TransactionMetadataEntry.TransactionMetadataOp.NEW)
+                        .setLastModificationTime(currentTimeMillis)
+                        .setMaxLocalTxnId(sequenceIdGenerator.getCurrentSequenceId());
+                if (owner != null) {
+                    transactionMetadataEntry.setOwner(owner);
+                }
```
If the owner is an empty string, should we set the owner to the transaction?

nicoloboschi replied:
I've added a check to ensure the owner is either null or not blank

codelipenghui commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2461,6 +2482,38 @@ protected void handleEndTxn(CommandEndTxn command) {
                 });
     }
 
+    private CompletableFuture<Boolean> verifyTxnOwnershipForBrokerToBrokerCommands() {
```
```suggestion
    private CompletableFuture<Boolean> verifyTxnOwnershipForTCToBrokerCommands() {
```

nicoloboschi replied:
done

codelipenghui commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2477,9 +2530,18 @@ protected void handleEndTxnOnPartition(CommandEndTxnOnPartition command) {
+                // we only accept super user becase this endpoint is reserved for broker to broker communication
+                verifyTxnOwnershipForBrokerToBrokerCommands()
+                        .thenCompose(isOwner -> {
+                            if (!isOwner) {
+                                return failedFutureTxnNotOwned(txnID);
```
It should not be the owner check, right? Essentially, it checks if the TC service is the super user.

liangyepianzhou replied:
+1.
In this case, we need to change the name of the method and modify the failure message.

liangyepianzhou replied:
And I have a question why we do not allow the transaction owner to end the transaction?
It is confused. A common user opens a transaction to do something, but he can not end the transaction since it's not a superuser.

nicoloboschi replied:
good point. 

the `endTxnOnPartition` and `endTxnOnSubscription` are targeted to the broker that owns the topic, so we might not be able to read the tnxMeta (the related TC might be loaded in another broker) so the owner check is not doable. I think it's fine, my assumption is that the broker client auth MUST match a super-user.

liangyepianzhou replied:
The `endTxnOnPartition` and `endTxnOnSubscription` are only called in `endTtansaction` which already has auth check, so I think the auth check for the superuser should be removed.

nicoloboschi replied:
A malicious attacker could still send this command impersonating the tc client, so we need to ensure that is allowed

liangyepianzhou replied:
You are right, thanks for your answer.

codelipenghui commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2561,23 +2622,29 @@ protected void handleEndTxnOnSubscription(CommandEndTxnOnSubscription command) {
+                // we only accept super user becase this endpoint is reserved for broker to broker communication
+                verifyTxnOwnershipForBrokerToBrokerCommands()
+                        .thenCompose(isOwner -> {
+                            if (!isOwner) {
+                                return failedFutureTxnNotOwned(txnID);
+                            }
```
Same as the above.

codelipenghui commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2658,9 +2726,15 @@ protected void handleAddSubscriptionToTxn(CommandAddSubscriptionToTxn command) {
-        transactionMetadataStoreService.addAckedPartitionToTxn(txnID,
-                MLTransactionMetadataStore.subscriptionToTxnSubscription(command.getSubscriptionsList()))
-                .whenComplete(((v, ex) -> {
+        verifyTxnOwnership(txnID)
```
Is it better to move the txn owner check to the transaction metadata store service?

nicoloboschi replied:
I moved only the part related to the interaction with txnMeta inside the txn metadata store service

codelipenghui commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2408,9 +2411,14 @@ protected void handleAddPartitionToTxn(CommandAddPartitionToTxn command) {
-        service.pulsar().getTransactionMetadataStoreService().addProducedPartitionToTxn(txnID,
-                command.getPartitionsList())
-                .whenComplete(((v, ex) -> {
+        verifyTxnOwnership(txnID)
```
Is it better to move to `TransactionMetadataStoreService`?

nicoloboschi replied:
done

liangyepianzhou commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2396,8 +2398,9 @@ protected void handleAddPartitionToTxn(CommandAddPartitionToTxn command) {
         final TxnID txnID = new TxnID(command.getTxnidMostBits(), command.getTxnidLeastBits());
         final TransactionCoordinatorID tcId = TransactionCoordinatorID.get(command.getTxnidMostBits());
         final long requestId = command.getRequestId();
+        final List<String> partitionsList = new ArrayList<>(command.getPartitionsList());
```
Why copy the data in the list again? Are there any special considerations?

nicoloboschi replied:
sorry it was not related, reverted

heesung-sohn committed:
[improve][broker] PIP-192 Added operation counters in ServiceUnitStateChannelImpl (#19410)

liangyepianzhou reviewed (approved):
LGTM

michaeljmarshall committed:
[cleanup][broker] Validate originalPrincipal earlier in ServerCnx (#19270)

congbobo184 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2408,9 +2409,15 @@ protected void handleAddPartitionToTxn(CommandAddPartitionToTxn command) {
-        service.pulsar().getTransactionMetadataStoreService().addProducedPartitionToTxn(txnID,
-                command.getPartitionsList())
-                .whenComplete(((v, ex) -> {
+        verifyTxnOwnership(txnID)
```
`addProducedPartitionToTxn` also invokes getTxnMeta, it is repeated with `verifyTxnOwnership`, how about moving it to TransactionMetadataStoreService and return `TxnMeta`.

nicoloboschi replied:
This would require to move the entire logic in the store implementation and pass the current role downstream. I don't think it's a good design because the ServerCnx contains the logic about the current session/role.

getting the TxnMeta is lightweight since it's in memory

congbobo184 replied:
Is it better if the verification happens after addProducedPartitionToTxn invoke `getTxnMeta `？make owner as `addProducedPartitionToTxn` parameter to TransactionMetadataStoreService

congbobo184 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2477,9 +2534,18 @@ protected void handleEndTxnOnPartition(CommandEndTxnOnPartition command) {
+                        })
                         .whenComplete((ignored, throwable) -> {
                             if (throwable != null) {
+                                throwable = handleTxnException(throwable,
```
we don't need to convert this exception, which is TC->TB, no need to handle CoordinatorNotFoundException

nicoloboschi replied:
fixed

congbobo184 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2561,23 +2626,29 @@ protected void handleEndTxnOnSubscription(CommandEndTxnOnSubscription command) {
+                            return subscription.endTxn(txnidMostBits, txnidLeastBits, txnAction, lowWaterMark);
+                        }).whenComplete((ignored, e) -> {
+                            if (e != null) {
+                                e = handleTxnException(e,
```
same as above

nicoloboschi replied:
fixed

michaeljmarshall committed:
[improve][test] Use channel.advanceTimeBy instead of Thread.sleep (#19447)

congbobo184 reviewed (commented):
LGTM! I think we should add a test for Tc recovered, then also can handle authentication.

nodece committed:
[fix][authorization] Fix the return value of canConsumeAsync (#19412)

Signed-off-by: Zixuan Liu <nodeces@gmail.com>

eolivelli reviewed (approved):
LGTM

nicoloboschi committed:
[fix][test] Fix flaky test: PendingAckPersistentTest.testDeleteUselessLogDataWhenSubCursorMoved (#19438)

michaeljmarshall committed:
[fix][broker] Expect msgs after server initiated CloseProducer (#19446)

tpiperatgod committed:
[improve][fn] support reading config options from file in Function Python Runner (#18951)

Signed-off-by: laminar <tpiperatgod@gmail.com>

Commit:
[improve][client] AuthenticationAthenz supports Copper Argos (#19445)

congbobo184 committed:
[improve][client] PIP-218: Consumer batchReceive() single partition every receive (#18316)

### Motivation
https://github.com/apache/pulsar/issues/18182

### Modifications

[<!-- Describe the modifications you've done. -->](https://github.com/apache/pulsar/issues/18182)

### Verifying this change
add test for batchReceive single topic messages
add test for batchReceive multi-topic messages
